### PR TITLE
Add  timestamp to agent commands

### PR DESCRIPTION
**Pull request by jcvenegas**
- Agent now can get log level configuration from cmdline.
- Changed logrus time format to : `TimestampFormat: time.RFC3339Nano` to get nanoseconds precision.
- Start and end of commands is logged with the message:   `{event}_start` or  `{event}_end`
- An special case is logged `agent_started seconds.nanoseconds` this get uptime with good clock precision.  
- Logrus format is JSON for easy parsing

### CODE REVIEW

File: agent.go
Comment (sboeuf) [inline agent.go]: You should use `defaultLogLevel` that you defined earlier.

File: agent.go
Comment (sboeuf) [inline agent.go]: Ditto.

File: agent.go
Comment (sboeuf) [inline agent.go:241]: Well that's not an error, it's about the proxy disconnecting, in order to end the agent.

File: network.go
Comment (sboeuf) [inline network.go]: That's an info, not a warning.

File: time.go
Comment (sboeuf) [inline time.go]: Add the Intel license header please.

**Review by sboeuf — CHANGES_REQUESTED** [review]

File: agent.go
Comment (jcvenegas) [reply agent.go:241]: I think the only way the proxy disconnecting is doing DESTROYPOD ?

File: agent_test.go
Comment (sboeuf) [inline agent_test.go]: s/kernel=someting/kernel=foo/

File: agent_test.go
Comment (sboeuf) [inline agent_test.go]: You could also test in case a wrong parameter is given. Something like:
`agent.debug=ino`
`agent.debug=wran`
`agent.debug=degbu`

File: agent_test.go
Comment (sboeuf) [inline agent_test.go]: Those messages don't help in the test validation, do they ?
I think we need something a bit more complex. You could create a pipe with `os.Pipe()`, and give one end of the pipe to logrus, and listen on the other end. That way for instance, you could check that you receive `info message` only when you enable at least Info level.

File: agent.go
Comment (sboeuf) [reply agent.go:241]: Ok fine.

File: agent.go
Comment (sboeuf) [inline agent.go]: Addressed in the wrong commit.

File: agent.go
Comment (sboeuf) [inline agent.go]: Ditto.

File: agent.go
Comment (sboeuf) [inline agent.go]: Use a constant for `"Failed to get uptime"` because you reuse it later.

File: agent.go
Comment (sboeuf) [inline agent.go]: Use a constant for the string.

File: agent.go
Comment (sboeuf) [inline agent.go]: Ditto

File: time_test.go
Comment (sboeuf) [inline time_test.go]: Add the Intel license header please.

**Review by sboeuf — CHANGES_REQUESTED** [review]

File: time.go
Comment (jodh-intel) [inline time.go]: This looks odd to me for 2 reasons:

- conventially, if a function return an `error`, you should ignore any other returned values, but you are creating a special "error" event.
- why can't the returned error be just `, err` rather than `, fmt.Errorf("%v", err)` ?

File: agent.go
Comment (jodh-intel) [inline agent.go]: It looks like the returned `error` won't include the event passed to `newEventTime()`. I'd create the string first:

```
event := hyper.CmdToString(cmd) + "_start"

if t, err := newEventTime(event); err == nil {
    agentLog.Infof("%s", t)
} else {
    agentLog.Errorf("Failed to get time for event %v: %v", event, err)
```

File: agent.go
Comment (jodh-intel) [inline agent.go]: I'd create a variable rather than calling `string()` twice:

```
str := string(w)
```

Also, I think you should check the number of fields returned by `Split()` before de-referencing the 2nd element.

File: time.go
Comment (jcvenegas) [reply time.go]: you are right!, fixed

File: agent_test.go
Comment (jcvenegas) [reply agent_test.go]: Indeed, it does not help, I added to help me know if was working but agree I can add the Pipe sugguestion

File: agent.go
Comment (MarioCarrilloA) [inline agent.go]: s/kenrel/kernel

File: agent.go
Comment (sboeuf) [inline agent.go]: Please edit the commit message and replace: `s/kenel/kernel/`

File: agent.go
Comment (sboeuf) [inline agent.go]: I am sorry but I don't like that we loop on agentOptions and at the same time we are modifying its values. Try something  more simple to stay as clear as possible:
I think we should expect `getConfig()` to return a config, and for each possible option parsed by `parseKernelCmdLine()`, we want a default in case it is not provided by the command line.
This means that `parseKernelCmdLine()` will build a list of options, that is returned to `getConfig()`, and from the `main()`, we apply this list of options in any case because we know that we will always have some default values (a function like `applyConfig()` would be fine). No need to specify any error, maybe just a `log.Warn()` in case we could read `/proc/cmdline` for example.

Moreover, about the way you loop over every element of this, because this is a map, you can directly have `key/value` from the loop for:
```go
for key, value := range optionList {
    ...
}
```

File: agent.go
Comment (sboeuf) [inline agent.go]: This should be part of the previous commit.

File: agent.go
Comment (sboeuf) [inline agent.go]: This commit message needs to be edited. `s/path/patch/` and `s/add/adds/`

**Review by sboeuf — CHANGES_REQUESTED** [review]

File: time.go
Comment (grahamwhaley) [inline time.go:28]: Hi @jcvenegas You asked me what time format I thought we should print in. My gut feeling is that we should probably use the same format as the other logs we produce within CC - so, for instance, it looks like the runtime logs use the logrus internal Time data:
https://github.com/clearcontainers/runtime/blob/9be2f54966b90014251afa5cebdf1ee36a85503f/logger.go#L137
If we can, I think we should do the same - what do you think @jodh-intel

Comment (jodh-intel) [reply time.go:28]: The actual timestamp itself should be in https://en.wikipedia.org/wiki/ISO_8601 format, or the RFC 3339 variant used by https://golang.org/pkg/time/. Looks like logrus uses that by default.

File: agent.go
Comment (jcvenegas) [reply agent.go]: Hi @sboeuf , I updated the PR:
1. I created agentConfig struct (by now we only have logLevel)
2. getConfig() return parse the kernel cmdline 
 - If could not parse an option I will use a default option.
3. applyConfig(config) will take a config and apply the values already parsed and validated in getConfig()

File: agent.go
Comment (sboeuf) [inline agent.go]: Don't assign the key that you parsed to the option, instead, use the key (that is `split[optionPosition]`) directly into the switch:
```go
switch split[optionPosition] {
    ...
```

File: agent.go
Comment (sboeuf) [inline agent.go]: You could add a default case here, and return an error like:
```go
default:
    return fmt.Errorf("Unknown option %s", split[optionPosition])
```

**Review by sboeuf — COMMENTED** [review]
Two comments, but other than that, the PR looks good to me now :)

File: agent.go
Comment (jodh-intel) [inline agent.go]: I think it would be better if this function also returned an `error`. The caller can then set `config.logLevel = defaultLogLevel`. It's neat to encapsulate the default handling like this, but it makes testing it difficult (how can you know which codepath was followed if `kernelCmdlineFile` contained the same `defaultLogLevel` value?)

Speaking of which, can you add a unit test for this function :smile:

**Review by sboeuf — APPROVED** [review]

**Comment by sboeuf** [comment]
LGTM
Let's wait for @jodh-intel comments before to merge it.

**Comment by jcvenegas** [comment]
@jodh-intel I still need to rework the timestamp format...

**Comment by jcvenegas** [comment]
@jodh-intel plase take a look to the log using logrus timestamp:
```
{"level":"info","msg":"Agent version: 0.1.0-alpha.0+-28cad9a9947ceced9a05b43ac3db5f4358ea5167","time":"2017-08-15T14:43:11.690803064Z"}
{"level":"info","msg":"agent_started 0.542392826","time":"2017-08-15T14:43:11.692726511Z"}
{"level":"info","msg":"Received \"startpod\" command","time":"2017-08-15T14:43:11.751076463Z"}
{"level":"info","msg":"startpod_start","time":"2017-08-15T14:43:11.751832257Z"}
{"level":"info","msg":"startpod_end","time":"2017-08-15T14:43:11.759136667Z"}
{"level":"info","msg":"Received \"newcontainer\" command","time":"2017-08-15T14:43:11.767398068Z"}
{"level":"info","msg":"newcontainer_start","time":"2017-08-15T14:43:11.767548997Z"}
```

**Comment by jodh-intel** [comment]
@jcvenegas - I'm not clear on your last comment. This looks fine to me.

However, are you planning to either modify the logrus formatter to change the time format, or maybe add a special time format *message* for `startpod_start` and `startpod_end` or something?

**Comment by jcvenegas** [comment]
@jodh-intel @grahamwhaley  I would like to get your approval in the format, now that I moved to RFC3339 as was suggested. 
- Yes I changed logrus time format to : `TimestampFormat: time.RFC3339Nano` to get nanoseconds precision.
- To log start end events I just print  `{event}_start` or  `{event}_end`
- An special case is logged `agent_started seconds.nanoseconds` this get uptime with good clock precision.  
- Logrus format is JSON for easy parsing

**Comment by sboeuf** [comment]
@jodh-intel @grahamwhaley could you give us your review on this one ?

**Comment by jodh-intel** [comment]
lgtm.

[Approved with PullApprove](https://pullapprove.com/clearcontainers/agent/pull-request/70/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/agent)

**Comment by grahamwhaley** [comment]
I'm happy if @jodh-intel is happy, and he is - so merging....

**Pull request merged**
